MSAL/src/instance/MSALDeviceInfoProvider.m

```diff
@@ -32,59 +32,91 @@
+                        completionBlock:(MSALDeviceInformationCompletionBlock)completionBlock
 {
-    if (![requestParameters shouldUseBroker])
-    {
-        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorBrokerNotAvailable, @"Broker is not enabled for this operation. Please make sure you have enabled broker support for your application", nil, nil, nil, nil, nil, YES);
-        completionBlock(nil, error);
-        return;
-    }
-    
-    
-    if (![MSIDSSOExtensionGetDeviceInfoRequest canPerformRequest])
-    {
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, requestParameters, @"Broker is not present on this device. Defaulting to personal mode");
-        
-        MSALDeviceInformation *msalDeviceInfo = [MSALDeviceInformation new];
-        msalDeviceInfo.deviceMode = MSALDeviceModeDefault;
-        completionBlock(msalDeviceInfo, nil);
-        return;
-    }
-    
-    NSError *requestError;
-    MSIDSSOExtensionGetDeviceInfoRequest *ssoExtensionRequest = [[MSIDSSOExtensionGetDeviceInfoRequest alloc] initWithRequestParameters:requestParameters
-                                                                                                                                  error:&requestError];
-    
-    if (!ssoExtensionRequest)
-    {
-        completionBlock(nil, requestError);
-        return;
-    }
-    
-    if (![self setCurrentSSOExtensionRequest:ssoExtensionRequest])
+    MSALDeviceInformation *deviceInformation = [MSALDeviceInformation new];
+
+    MSIDRegistrationInformation* regInfo = [MSIDWorkPlaceJoinUtil getRegistrationInformation:nil urlChallenge:nil];
+    if (regInfo && regInfo.isWorkPlaceJoined)
     {
-        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Trying to start a get accounts request while one is already executing", nil, nil, nil, nil, nil, YES);
-        completionBlock(nil, error);
-        return;
+        // Certificate subject is nothing but the AAD deviceID
+        NSString *deviceId = regInfo.certificateSubject;
+        if (deviceId)
+        {
+            [deviceInformation addWorkPlaceJoinedDeviceId:deviceId];
+        }
+
+        NSString *upn = [MSIDWorkPlaceJoinUtil getWPJStringData:nil identifier:kMSIDUPNKeyIdentifier error:nil];
+        if (upn)
+        {
+            [deviceInformation addWorkPlaceJoinedUPN:upn];
+        }
+
+        NSString *tenantID = [MSIDWorkPlaceJoinUtil getWPJStringData:nil identifier:kMSIDTenantKeyIdentifier error:nil];
+        if (tenantID)
+        {
+            [deviceInformation addWorkPlaceJoinedTenantId:tenantID];
+        }
     }
-    
-    [ssoExtensionRequest executeRequestWithCompletion:^(MSIDDeviceInfo * _Nullable deviceInfo, NSError * _Nullable error)
+
+    if (@available(iOS 13.0, macOS 10.15, *))
     {
-        [self copyAndClearCurrentSSOExtensionRequest];
-        
-        if (!deviceInfo)
+        if (![requestParameters shouldUseBroker])
```

Comment:
this check is still needed for all cases. We should only support WPJ operations for supported scenarios.

Reply:
Which check "if (@available(iOS 13.0, macOS 10.15, *))" or "if (![requestParameters shouldUseBroker])" ?

Reply:
Ok so you mean to say, we should returns WPJ only of requestParameters contains shouldUseBroker true?

Reply:
Yes